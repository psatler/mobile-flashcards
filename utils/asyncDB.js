import { AsyncStorage } from 'react-native'
import { data } from './mockData'
const FLASHCARDS_DECKS_KEY = 'Flashcards:decks'

const setInitialData = async () => {
    let obj = data;

    await AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(obj))
}


export const getDecks = async () => {
    try {
        let value = await AsyncStorage.getItem(FLASHCARDS_DECKS_KEY);
        if( value !== null){ //we have data
            // console.log('decks', value);
            let parsedObj = JSON.parse(value); //parsing back to object
            return parsedObj;
        } else { //if not, we set initial state
            console.log('setting initial state');
            setInitialData(); //defined above
        }
    } catch (err) {
        console.log('There was an error retrieving decks from local DB: ', err);
    }
}

export const saveDeckTitle = (keyTitle, deckObj) => {
    try {
        return AsyncStorage.mergeItem(FLASHCARDS_DECKS_KEY, JSON.stringify({
            [keyTitle]: deckObj
        }))
        
    } catch (err) {
        console.log(err);
    }
}

export const addCardToDeck = (deckTitle, card) => {
    const key = deckTitle;

    return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
        .then( (results) => {
            const allDecks = JSON.parse(results);
            const deck = allDecks[key];
            // console.log('addCardToDeck- BEFORE', deck.questions)
            deck.questions.push(card)
            // console.log('addCardToDeck', deck.questions)
            // console.log('allDecks', allDecks)

            AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(allDecks)); //updating 

        })
        // const key = deckTitle.split(' ').join('');

}

// 

// const checkIfDataExists = (results) => {
//     return results === null
//       ? setInitialData()
//       : JSON.parse(results)
//   }

// export const retrieveDataFromAsyncStorage = () => {

//     return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
//         .then(checkIfDataExists)
// }



// _setValue = async () => {
//     try {
//       await AsyncStorage.setItem(STORAGE_KEY, 'true');
//     } catch (error) { // log the error
//     }
//   };

//   _loadInitialState = async () => {
//     try {
//       let value = await AsyncStorage.getItem(STORAGE_KEY);
//       if (value === 'true'){
//         this.setState({loaded: 'true'});
//       } else {
//         this.setState({loaded: 'false'});
//         this._setValue();
//       }
//     } catch (error) {
//       this.setState({loaded: 'false'});
//       this._setValue();
//     }
//   };