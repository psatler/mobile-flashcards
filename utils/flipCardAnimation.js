import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Animated
} from 'react-native';

export default class AnimatedBasic extends Component {

    state = {
        animatedValue: new Animated.Value(0),
        listenerValue: 0,
    }

    componentDidMount(){
        const { animatedValue, listenerValue } = this.state;

        //creating a listener to track values changes
        animatedValue.addListener( ( { value } ) => {
            this.setState({
                listenerValue: value,
            })
        })
    }

    flipCard = () => {
        const { animatedValue, listenerValue } = this.state;

        if(listenerValue >= 90){
            Animated.spring(animatedValue,{
                toValue: 0,
                friction: 8,
                tension: 10
              }).start();
        } 
        else {
            Animated.spring(animatedValue,{
                toValue: 180,
                friction: 8,
                tension: 10
              }).start();
        }
    }

    render() {
        const { animatedValue, listenerValue } = this.state;

        const frontAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
                outputRange: ['0deg', '180deg'],
              })
            }
          ],
          //adding opacity so the back face won't be mirrored on Android devices
        //https://github.com/facebook/react-native/issues/1973#issuecomment-262059217
          opacity: animatedValue.interpolate({ //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90],
                outputRange: [1, 0]
            }), 
        }
        const backAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
              //   outputRange: ['180deg', '360deg'],
                outputRange: ['180deg', '0deg']
              }) 
            }
          ],
          opacity: animatedValue.interpolate({  //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90], 
                outputRange: [0, 1] 
            }), 
        }
        
        return (
          <View style={styles.container}>
            <View>
              <Animated.View style={[styles.flipCard, frontAnimatedStyle]}>
                <Text style={styles.flipText}>
                  This text is flipping on the front.
                </Text>
              </Animated.View>
              <Animated.View style={[backAnimatedStyle, styles.flipCard, styles.flipCardBack]}>
                <Text style={styles.flipText}>
                  This text is flipping on the back.
                </Text>
              </Animated.View>
            </View>
            <TouchableOpacity onPress={() => this.flipCard()}>
              <Text>Flip!</Text>
            </TouchableOpacity>
          </View>
        );
      }
    }


const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  flipCard: {
    width: 200,
    height: 200,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'blue',
    backfaceVisibility: 'hidden', //it means when the card rotates out of the way, it won't be visible. When it rotates back, it'll be visible again. Not working on Android though, so it was added an workaround for this above
  },
  flipCardBack: {
    backgroundColor: "red",
    position: "absolute",
    top: 0,
  },
  flipText: {
    width: 90,
    fontSize: 20,
    color: 'white',
    fontWeight: 'bold',
  }
});