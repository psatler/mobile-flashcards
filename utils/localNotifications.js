// import React from 'react'
import { AsyncStorage } from 'react-native'
import { Notifications, Permissions } from 'expo'

const FLASHCARDS_NOTIFICATIONS_KEY = 'Flashcards:localNotifications'


const createNotification = () => {
    return {
        title: 'Do a quiz today!',
        body: 'ðŸ‘‹ Do not forget to study your decks!',
        ios: {
            sound: true,
        },
        android: {
            sound: true,
            priority: 'high',
            sticky: false,
            vibrate: true,
        }
    }
}

export const setLocalNotification = () => {
    //checking if our local notification has already been set
    AsyncStorage.getItem(FLASHCARDS_NOTIFICATIONS_KEY) 
        .then(JSON.parse) //parsing the returned value
        .then( (data) => {
            if(data === null){ //if we haven't any notification scheduled
                Permissions.askAsync(Permissions.NOTIFICATIONS)
                    .then( ({ status }) => {
                        if( status === 'granted' ){
                            Notifications.cancelAllScheduledNotificationsAsync();

                            let tomorrow = new Date();
                            tomorrow.setDate(tomorrow.getDate() + 1) //tomorrow 
                            tomorrow.setHours(10); //scheduling for 10 AM
                            tomorrow.setMinutes(0);

                            Notifications.scheduleLocalNotificationAsync(
                                createNotification(),
                                {
                                    time: tomorrow,
                                    repeat: 'day'
                                }
                            )

                            AsyncStorage.setItem(FLASHCARDS_NOTIFICATIONS_KEY, JSON.stringify(true))
                        }
                    })

            }
        })

}

export const clearLocalNotification = () => {
    return AsyncStorage.removeItem(FLASHCARDS_NOTIFICATIONS_KEY)
      .then(Notifications.cancelAllScheduledNotificationsAsync)
}