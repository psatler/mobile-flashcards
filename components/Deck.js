import React from 'react'
import { Text, View, StyleSheet, Platform, Image } from 'react-native'
import { lightBlue } from '../utils/colors';
import { Entypo } from '@expo/vector-icons'

const AVATAR_IMAGE_HEIGHT = 40

const Deck = (props) => {
    return (
        <View style={deckContainer}>
            <View style={ styles.avatar} >

                <Image 
                    source={props.imageURI ? {uri: props.imageURI } : require('../assets/reactNativeWhiteBackground.png') }
                    style={styles.image}
                />
            </View>
            <View style={ styles.deckInfo} >
                <Text style={deckTitle} >{props.deckName}</Text>
                <Text style={deckSize} >{props.deckSize} cards</Text>
            </View>
            <View style={ styles.removeButton} >
                <Entypo name='dots-three-vertical' size={30}  />
            </View>
        </View>
    )
}

export default Deck;


const styles = StyleSheet.create({
    deckContainer: {
        // flex: 1, //uncomment this so in the details screen it takes the whole screen space
        flexDirection: 'row',
        margin: 5,
        padding: 5,
        borderRadius: Platform.OS === 'ios' ? 16 : 2,
        borderColor: lightBlue,
        borderWidth: 1,
        minHeight: 80,
        // backgroundColor: white,
    },
    image: {
        flex: 1,
        width: null,
        height: null,
    },
    avatar: {
        flex: 0.5,
        borderRadius: AVATAR_IMAGE_HEIGHT/2, 
        borderColor: lightBlue,
        borderWidth: 1,
        overflow: 'hidden'
    },
    deckInfo: {
        flex: 2,
        alignItems: 'center',
    },
    deckTitle: {
        fontWeight: 'bold',
        fontSize: 20,
    },
    deckSize: {
        marginTop: 10,
        fontWeight: 'bold',
        fontSize: 15,
    }, 
    removeButton: { 
        flex: 0.5, 
        // backgroundColor: purple,
        justifyContent: 'center',
        alignItems: 'center',
    }
})
const { deckContainer, deckTitle, deckSize } = styles;