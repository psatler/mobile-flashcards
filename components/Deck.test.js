import React from 'react';
import Deck from './Deck';
import { data } from '../utils/mockData'

import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';


describe('[Component] Deck', () => {
    const singleDeck = data['JavaScript'];

    it('renders correctly', () => {
        const rendered = renderer.create(
            <Deck 
                deckName={singleDeck.title}
                deckSize={singleDeck.questions.length}
                imageURI={singleDeck.image}
            />).toJSON();
        expect(rendered).toMatchSnapshot();
    });

    it('renders user shallow correctly', () => {
        const renderer = new ShallowRenderer();
        renderer.render(
            <Deck 
                deckName={singleDeck.title}
                deckSize={singleDeck.questions.length}
                imageURI={singleDeck.image}
            />);
        const tree = renderer.getRenderOutput();
        expect(tree).toMatchSnapshot();
      });
});


/**
 * $ 
 *  npm install --save-dev react-dom react-addons-test-util
    npm WARN notice Due to a recent security incident, all user tokens have been invalidated. Please see https://status.npmjs.org/incidents/dn7c1fgrr7ng for more details. To generate a new token, visit https://www.npmjs.com/settings/~/tokens or run "npm login".
    npm ERR! code E404
    npm ERR! 404 Not Found: react-addons-test-util@latest

    npm ERR! A complete log of this run can be found in:
    npm ERR!     C:\Users\Pablo\AppData\Roaming\npm-cache\_logs\2018-07-19T21_16_41_344Z-debug.log
 */
