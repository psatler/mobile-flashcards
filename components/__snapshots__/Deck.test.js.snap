// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Deck renders correctly 1`] = `
<View
  style={
    Object {
      "borderColor": "#00A4D3",
      "borderRadius": 16,
      "borderWidth": 1,
      "flexDirection": "row",
      "margin": 5,
      "minHeight": 80,
      "padding": 5,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#00A4D3",
        "borderRadius": 20,
        "borderWidth": 1,
        "flex": 0.5,
        "overflow": "hidden",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "flex": 1,
          "height": null,
          "width": null,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 2,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      JavaScript
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 15,
          "fontWeight": "bold",
          "marginTop": 10,
        }
      }
    >
      1
       cards
    </Text>
  </View>
</View>
`;

exports[`[Component] Deck renders user shallow correctly 1`] = `
<Component
  style={
    Object {
      "borderColor": "#00A4D3",
      "borderRadius": 16,
      "borderWidth": 1,
      "flexDirection": "row",
      "margin": 5,
      "minHeight": 80,
      "padding": 5,
    }
  }
>
  <Component
    style={
      Object {
        "borderColor": "#00A4D3",
        "borderRadius": 20,
        "borderWidth": 1,
        "flex": 0.5,
        "overflow": "hidden",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "flex": 1,
          "height": null,
          "width": null,
        }
      }
    />
  </Component>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flex": 2,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      JavaScript
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 15,
          "fontWeight": "bold",
          "marginTop": 10,
        }
      }
    >
      1
       cards
    </Text>
  </Component>
</Component>
`;
